# EAPhy
Exon Alignment for Phylogenetics

EAPhy, consists of a collection of scripts that jointly process exonic sequences, use an existing aligner, check and filter alignments and ultimately generates the required input for commonly used phylogenetic inference programs. The main aim of this pipeline is to provide a flexible but rigorous tool to generate reliable alignments and by minimising the time to do so, promote the application of multiple species tree methods and examine the (in)congruences between distinct inference approaches.

All scripts are written in Python and throughout the pipeline, many functions benefit from modules that have been developed as part of the BioPython distribution (http:// biopython.org/wiki/Main_Page).

Feel free to use EAPhy or any of its associated scripts, but please acknowledge the great work of the BioPython team and the communication describing EAPhy (Blom MPK, 2015) in any forthcoming publications.

Please contact me if any issues arise or you have any suggestions/requests.

Mozes Blom
mozes.blom@gmail.com

Blom MPK. EAPhy: A Flexible Tool for High-throughput Quality Filtering of Exon-alignments and Data Processing for Phylogenetic Methods. PLOS Currents Tree of Life. 2015 Aug 5 . Edition 1. doi: 10.1371/currents.tol.75134257bd389c04bc1d26d42aa9089f.
